pmd-plsql/etc/grammar/PldocAST.jjt

```diff
@@ -2494,57 +2494,68 @@ ASTEqualityExpression EqualityExpression() #EqualityExpression(>1) :
  }
 }
 
+/**
+ * Relational operators: https://docs.oracle.com/en/database/oracle/oracle-database/18/lnpls/plsql-language-fundamentals.html#GUID-2EB964C8-9B36-4009-A45F-B340F44EE2A2
+ */
 ASTRelationalExpression RelationalExpression() #RelationalExpression(>1) :
 { PLSQLNode simpleNode = null; StringBuilder sb = new StringBuilder() ; }
 {
  (
   //AdditiveExpression() ( ( "<" | ">" | "<=" | ">=" | [<NOT>] (<IN> | <BETWEEN> | <LIKE>) ) AdditiveExpression() )*
-  (simpleNode = AdditiveExpression() ) { sb.append(simpleNode.getImage()); }
-  ( 
-   ( 
+  ( simpleNode = AdditiveExpression() ) { sb.append(simpleNode.getImage()); }
+  (
+    (
+
       (
-	(
-	    (("<" ) { sb.append("<"); } 
-	     [">"  { sb.append(">"); } ] // <> 
-	    )
-	  | (">" ) { sb.append(">"); }
-	)
-	[ ("=" ) { sb.append("="); } ]
-    
+          ("<>" { sb.append(" <> "); } )
+        | ("<=" { sb.append(" <= "); } )
+        | (">=" { sb.append(" >= "); } )
+        | ("!=" { sb.append(" != "); } )
+        | ("~=" { sb.append(" ~= "); } )
+        | ("^=" { sb.append(" ^= "); } )
+        | ("<" { sb.append(" < "); } )
+        | (">" { sb.append(" > "); } )
       )
+
       { sb.append(" "); } 
-   | [(<NOT>) { sb.append(" NOT "); }] 
-     ((<IN>) { sb.append(" IN "); } 
-     | (<BETWEEN> ) { sb.append(" BETWEEN "); }
-     | (<LIKE> ) { sb.append(" LIKE "); }
-     | (
-	 (
-	   (<MEMBER>) { sb.append(" MEMBER "); }
-	   |
-	   (<SUBMULTISET>) { sb.append(" SUBMULTISET "); }
-	 )
-         [ <OF> { sb.append(" OF "); } ]
-       )
-     ) 
-    // MULTISET Operators 
-    | ( 
-       <MULTISET> { sb.append (" MULTISET " );}  
-       (   <EXCEPT>  { sb.append (" EXCEPT " );}  
-         | <INTERSECT>  { sb.append (" INTERSECT " );} 
-	 | <UNION>  { sb.append (" UNION " ); }    
-       )
-       [   <DISTINCT>  { sb.append (" DISTINCT " );} 
-         | <ALL>  { sb.append (" ALL " );}   
-       ]
-     ) 
-   ) 
-   (AdditiveExpression() ) { sb.append(simpleNode.getImage()); }
-   [
-	<ESCAPE> { sb.append(" ESCAPE ");} 
-	(  <CHARACTER_LITERAL>{ sb.append(" "); sb.append(token.toString()); }  
-	  | simpleNode = StringLiteral() { sb.append(" "); sb.append(simpleNode.getImage()); }  
-	)
-   ]
+      | [(<NOT>) { sb.append(" NOT "); }]
```

Comment:
Doesn't this production try to do too much? I mean dumping the whole operator to a string makes it harder to analyse later on, whereas reifying the structure of the operator into an AST node would avoid losing this information.
E.g.
```
RelationalExpression ::= 
    AdditiveExpression RelationalOperator AdditiveExpression EscapeClause?

RelationalOperator  ::= // a #void interface, to keep the AST flat
      ComparisonOperator
    | NegatedOperator
    | NegatableOperator

NegatedOperator ::=
   "NOT" NegatableOperator
     
NegatableOperator ::= // a #void interface
      BetweenOp   // These nodes would each correspond to their respective keyword
    | InOp
    | LikeOp
    | SetInclusionOp
    | MultisetOp

ComparisonOperator ::=
    "<>" | "<=" | ...

SetInclusionOp ::= 
    (MemberOp | SubmultisetOp) "OF"?

MultisetOp ::=
   "MULTISET" ("EXCEPT | "INTERSECT" | "UNION") [ "DISTINCT" | "ALL"]

EscapeClause ::=
  "ESCAPE" (CharLiteral | StringLiteral)

```

Having one node for each operator like
``` 
    | InOp
    | LikeOp
```
may seem verbose at first, but JJTree allows to write that inline as e.g.
```java
      <IN>   #InOp
    | <LIKE> #LikeOp
``` 
instead of splitting this in very many productions. (I mention this because I think I've never come across this syntax anywhere in our grammars so maybe you're unaware of it.) This makes the AST structure very informative, but has the downside to increase the number of node classes significantly. If this is a problem, we may want to use a single common node type, e.g. `ASTPlsqlKeyword`, to represent any one-keyword AST node, with the image selecting the actual keyword used.

This is obviously out of the scope of this PR, and maybe to be left for 7.0.0, but something to think about anyway.
I can open an issue to track it separately if you think this change is worth it.

Reply:
Writing the grammar the way you propose, is definitively a better way of doing it. I've created #1500 for this.